crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder.rs

```diff
@@ -0,0 +1,528 @@
+            to,
+            FactorSelector::Kind(kind),
+            quantity_to_add,
+        )
+    }
+
+    fn assign_one_hardware_factor(&mut self, to: RoleKind) {
+        self.assign_factors_of_category(
+            to,
+            FactorSourceCategory::Hardware,
+            Quantity::one(),
+        )
+    }
+
+    fn count_factors_for_role(&self, role_kind: RoleKind) -> u8 {
+        self.factors_for_role(role_kind).len() as u8
+    }
+
+    fn assign_factors_of_category_to_recovery(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_of_category(Recovery, category, quantity_to_add)
+    }
+
+    fn assign_factors_of_category_to_confirmation(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) {
+        self.assign_factors_of_category(Confirmation, category, quantity_to_add)
+    }
+
+    /// Automatic assignment of factors to roles according to [this heuristics][doc].
+    ///
+    /// [doc]: https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3758063620/MFA+Rules+for+Factors+and+Security+Shields#Automatic-Security-Shield-Construction
+    fn assign(&mut self) -> Result<ProtoMatrix> {
+        // 📒 "If the user only chose 1 factor for PRIMARY, remove that factor from the list (it cannot be used elsewhere - otherwise it can)."
+        {
+            if self.count_factors_for_role(Primary) == 1
+                && let Some(only_primary_factor) =
+                    self.proto_matrix.primary.iter().next()
+            {
+                self.remaining_available_factors
+                    .retain(|f| f.id() != *only_primary_factor);
+            }
+        }
+
+        // 📒 "Drop in the somewhat “special-use” factors first"
+        {
+            // 📒	"Add all Contact factors in the list to RECOVERY."
+            self.assign_factors_of_category_to_recovery(Contact, Quantity::All);
+
+            // 📒	"Add all Information factors in the list to CONFIRMATION."
+            self.assign_factors_of_category_to_confirmation(
+                Information,
+                Quantity::All,
+            );
+        }
+
+        let mut distribute_custodian_and_hardware_to_non_primary = || {
+            // 📒 "Add any Custodian.." 🙅‍♀️  Custodian FactorSources does not exist yet...
+
+            // 📒 "Add any Hardware factors in the list, starting with the most recently used, to RECOVERY
+            // until there is at least _ factor source in RECOVERY."
+            self.assign_one_hardware_factor(Recovery);
+
+            // 📒 "Add any Hardware (Ledger, Arculus, Yubikey) factors in the list, starting with the most recently used, to CONFIRMATION until there is at least _ factor factors in CONFIRMATION."
+            self.assign_one_hardware_factor(Confirmation);
+        };
+
+        // 📒 "Distribute to try to get at least 1 RECOVERY and then 1 CONFIRMATION" [if possible]
```

Comment:
@ymatto specifically it is this part I want feedback on. Here I try to first add a Hardware factor to `Recovery` and then one to `Confirmation`, and then same again. **Unconditionally**, and it is OK if non was found.

Did you want to make this conditional? As in, only add if number of factors for that role is equal, less than or greater than some threshold? I think this is correct... just that the text in confluence `until there is at least` confused me...